### PR TITLE
Packets working branch

### PR DESCRIPTION
**Pull request by CollinValley**
This is a huge commit, everyone please go through it and point out anything that is hacky, could use improvement. Particularly, I want to weed out things that likely don't do what you expect. I want this library to not easily panic, and have a what you see is what you get feel to it.  This is a big ol' boy, and I'm going to want to merge it not completely perfect, but I think it is in a functional enough state now that it could be used to begin building elements.  Once we have a list of fixes etc, I'll pare down and clean up the commit messages.

### CODE REVIEW

File: route-rs-runtime/src/lib.rs
Comment (moosingin3space) [inline route-rs-runtime/src/lib.rs]: I think this could work better as a standalone crate -- that way I could use it in `afpacket` tests, for instance.

File: route-rs-runtime/src/packet/ipv6.rs
Comment (CollinValley) [inline route-rs-runtime/src/packet/ipv6.rs]: Forgot to switch this so the new function takes the variable packet_offset length. Will get that tomorrow.

Comment (CollinValley) [reply route-rs-runtime/src/packet/ipv6.rs]: See ipv4::new() for the idea.

File: route-rs-runtime/src/packet/ethernet.rs
Comment (moosingin3space) [inline route-rs-runtime/src/packet/ethernet.rs]: Let's think about a legit `Error` type. I'm not too sure what the best way to do `Error`s is since `failure` became superseded by the stdlib though.

File: route-rs-runtime/src/packet/ethernet.rs
Comment (moosingin3space) [inline route-rs-runtime/src/packet/ethernet.rs]: If `MacAddr` stored an array internally, you could use [this](https://doc.rust-lang.org/std/primitive.slice.html#method.copy_from_slice).

File: route-rs-runtime/src/packet/ethernet.rs
Comment (moosingin3space) [inline route-rs-runtime/src/packet/ethernet.rs]: Any reason for `From` instead of `TryFrom` here?

File: route-rs-runtime/src/packet/ipv4.rs
Comment (moosingin3space) [inline route-rs-runtime/src/packet/ipv4.rs]: Ideally a packet type doesn't have to worry about what its containing packet type is. This is worth discussion.

File: route-rs-runtime/src/packet/packet.rs
Comment (moosingin3space) [inline route-rs-runtime/src/packet/packet.rs]: See above comments about `copy_from_slice`.

File: route-rs-runtime/src/packet/udp.rs
Comment (moosingin3space) [inline route-rs-runtime/src/packet/udp.rs]: Not sure "segment" is the semantically-correct name for this -- `UdpDatagram` maybe?

**Review by moosingin3space — CHANGES_REQUESTED** [review]
I like this API. That said, it'd be much easier to read with constants, and I'd like to have a discussion about ways we can potentially make each inner packet independent of the outer packet storing it.

File: route-rs-runtime/src/packet/ethernet.rs
Comment (CollinValley) [reply route-rs-runtime/src/packet/ethernet.rs]: Yeah, I would like to be able to return formatted strings or something.

File: route-rs-runtime/src/packet/udp.rs
Comment (CollinValley) [reply route-rs-runtime/src/packet/udp.rs]: In TCP they are called Segments, UDP, doesn't seem like it has a defined term. Some people say segment, others say packet.

File: route-rs-runtime/src/packet/ipv4.rs
Comment (CollinValley) [reply route-rs-runtime/src/packet/ipv4.rs]: This is an old commit. In the latest I added a packet_offset parameter that means the packet can be agnostic to the amount of data preceeding it in the array.

File: route-rs-runtime/src/packet/ethernet.rs
Comment (CollinValley) [reply route-rs-runtime/src/packet/ethernet.rs]: /facepalm this is exactly what I wanted when I made the Result.  https://doc.rust-lang.org/std/convert/trait.TryFrom.html Will convert that over tonight.

File: route-rs-runtime/src/packet/ethernet.rs
Comment (CollinValley) [reply route-rs-runtime/src/packet/ethernet.rs]: Yeah, I will switch it over, similarly to how I did Ipv4/Ipv6. They keep an array internally.

File: route-rs-runtime/src/packet/ethernet.rs
Comment (CollinValley) [reply route-rs-runtime/src/packet/ethernet.rs]: I'll take a look into Error types. Ideally we could return formatted strings

**Comment by CollinValley** [comment]
> I like this API. That said, it'd be much easier to read with constants, and I'd like to have a discussion about ways we can potentially make each inner packet independent of the outer packet storing it.

Yeah,  I made some strides on this front by retaining offsets as we go up stack.  A TCPsegment keeps track of the IP Header offset, it's own TCP Header offset, an then it's payload offset. This allows us to remain more agnostic about why layer is beneath us. Unfortunately, there are instances of coupling, like the combined checksum in TCP, that break the model.  Also the fact that TCP has no length field, it just uses the IP length. Super annoying.

File: route-rs-runtime/src/lib.rs
Comment (CollinValley) [reply route-rs-runtime/src/lib.rs]: I want to agree, would be nice to version the library too, so I can tell users it is 0.1 lol.  Will create a patch to move the files.

File: route-rs-runtime/src/packet/types.rs
Comment (moosingin3space) [inline route-rs-runtime/src/packet/types.rs]: Switch to single colon.

File: route-rs-runtime/src/packet/types.rs
Comment (moosingin3space) [inline route-rs-runtime/src/packet/types.rs]: This already exists in the stdlib.

File: route-rs-runtime/src/packet/types.rs
Comment (moosingin3space) [inline route-rs-runtime/src/packet/types.rs]: See above.

File: route-rs-runtime/src/packet/ipv6.rs
Comment (moosingin3space) [inline route-rs-runtime/src/packet/ipv6.rs]: Can you cite your source here please?

File: route-rs-runtime/src/packet/ipv6.rs
Comment (moosingin3space) [inline route-rs-runtime/src/packet/ipv6.rs]: Use a common function (between this and `get_ipv6_payload_type`)

**Review by moosingin3space — DISMISSED** [review]

File: route-rs-runtime/src/packet/ipv6.rs
Comment (CollinValley) [reply route-rs-runtime/src/packet/ipv6.rs]: Took at look at this, while a good amount of the logic is the same. The main loop differs which makes abstracting the logic hard.

File: route-rs-runtime/src/packet/types.rs
Comment (CollinValley) [reply route-rs-runtime/src/packet/types.rs]: Done

**Comment by CollinValley** [comment]
I rebased this against master, so it once I have a little bit of approval we can merge this in. I think it is in a useful state right now, enough that I can call it version 0.1 It should also unblock us from building a router/elements using it.

File: route-rs-packets/src/ipv4.rs
Comment (eeverett6) [inline route-rs-packets/src/ipv4.rs:225]: Why not an error type? What if this reserved value is later assigned?

File: route-rs-packets/src/ipv6.rs
Comment (eeverett6) [inline route-rs-packets/src/ipv6.rs:184]: Same question about reserved here

File: route-rs-packets/src/types.rs
Comment (eeverett6) [inline route-rs-packets/src/types.rs:197]: Couple thoughts:
- It feels wordy to have to do this twice
- How do we turn an IpProtocol into a u8 when writing out a packet/overwriting the protocol field?

Have you seen the solution in libpnet? https://github.com/libpnet/libpnet/blob/master/pnet_packet/src/ip.rs It looks like the protocol is a struct with a single u8 field, then there are two mappings, one a list of constants, and one a display function.

File: route-rs-packets/src/ethernet.rs
Comment (eeverett6) [inline route-rs-packets/src/ethernet.rs]: In Ethernet II, the 12th and 13th bytes are actually the ethertype field, not a payload length. I believe there is no length field in ethernet II. We will probably need something like the IpProtocol types for EtherTypes as well

File: route-rs-packets/src/ipv4.rs
Comment (eeverett6) [inline route-rs-packets/src/ipv4.rs]: dscp

File: route-rs-packets/src/ethernet.rs
Comment (eeverett6) [inline route-rs-packets/src/ethernet.rs:80]: Should we have a way of encapsulating an Ipv4 packet if it didn't originally come from an ethernet frame?

**Review by eeverett6 — COMMENTED** [review]
Didn't have time to look over the whole thing tonight, but posting my feedback so far. I should be able to contribute commits to address some of this feedback myself soon too.

File: route-rs-packets/src/ipv4.rs
Comment (CollinValley) [reply route-rs-packets/src/ipv4.rs:225]: Idk making an error type is kind of annoying, this is quick and dirty. Error type is on the list of improvements, i can spawn off an issue.

Comment (CollinValley) [reply route-rs-packets/src/ipv4.rs:225]: I can fix this by returning a Result at least to start.

File: route-rs-packets/src/ipv6.rs
Comment (CollinValley) [reply route-rs-packets/src/ipv6.rs:184]: ack

File: route-rs-packets/src/types.rs
Comment (CollinValley) [reply route-rs-packets/src/types.rs:197]: Hmm this is a decent solution.  I'd be fine with changing it to something like libpnet's

File: route-rs-packets/src/ipv4.rs
Comment (CollinValley) [reply route-rs-packets/src/ipv4.rs]: ack

File: route-rs-packets/src/ethernet.rs
Comment (CollinValley) [reply route-rs-packets/src/ethernet.rs:80]: In this case, this function creates an EthernetFrame from an Ipv4 or Ipv6.  So it isn't really encapsulating per se, just a type conversion.  We could add support for other Layer 2 protocols if need be, since the try_from would fail in that caseif you tried to make an ethernet frame.

File: route-rs-packets/src/ethernet.rs
Comment (CollinValley) [reply route-rs-packets/src/ethernet.rs]: Interesting.  https://en.wikipedia.org/wiki/Ethernet_frame Looks like it is EtherType in the case it is an Ethernet II frame. As far as I'm aware, we really only expect Ethernet II types on the internet?  https://en.wikipedia.org/wiki/EtherType Looks like it is fairly complex.

**Comment by CollinValley** [comment]
Thanks for looking through it Ethan, obviously any contribution is welcome.  I'll start trying to address the simple things, and will create some issues for the stuff that will take longer.

File: route-rs-packets/src/ipv4.rs
Comment (moosingin3space) [reply route-rs-packets/src/ipv4.rs:225]: What kind of errors are possible?

File: route-rs-packets/src/types.rs
Comment (moosingin3space) [reply route-rs-packets/src/types.rs:197]: Yeah, arguably `From<T>` isn't what we want here -- let's make clients explicitly convert integers to the typed values.

File: route-rs-packets/src/ethernet.rs
Comment (moosingin3space) [reply route-rs-packets/src/ethernet.rs]: EtherTypes are how we define what the next packet type is, though -- since LLDP and ARP don't use IP headers. I'm not sure we can avoid it. According to https://en.wikipedia.org/wiki/Ethernet_frame#Types, it looks like Ethernet II is the most common, with raw IEEE 802.3 being an ancient Novell thing.

File: route-rs-packets/src/types.rs
Comment (CollinValley) [reply route-rs-packets/src/types.rs:197]: Going to shelve this for after merge cleanup.

Comment (CollinValley) [reply route-rs-packets/src/types.rs:197]: #178

File: route-rs-packets/src/ipv4.rs
Comment (CollinValley) [reply route-rs-packets/src/ipv4.rs]: There are a few numbers that are undefined?

Comment (CollinValley) [reply route-rs-packets/src/ipv4.rs:225]: Edit: Nevermind this error makes sense. The only error is that they didn't hand me something that could have been an Ipv4 packet.

I am going to wrap this in a Result rather than a custom error to be more clear.

**Comment by CollinValley** [comment]
Just resolved the last issue from the previous CRs. Also rebased the branch against master to ensure everything is still working as expected.

File: route-rs-packets/src/ethernet.rs
Comment (eeverett6) [reply route-rs-packets/src/ethernet.rs]: I would honestly just drop support for the 802.3 stuff entirely and just support Ethernet II with the EtherType field. I don't think it's worth it to be generic across so many different standards when we'll never encounter them.

File: route-rs-packets/src/ethernet.rs
Comment (eeverett6) [reply route-rs-packets/src/ethernet.rs:80]: I'm thinking of the case where say a processor/element creates an Arp Message, that now needs to be put behind an Ethernet header to be sent out over the wire. These would be separate from the TryFroms, but could be methods that take an arbitrary payload, a src/dest MAC addr, and an ethertype, then replaces the old l2 header or pushes more data onto the front of the packet as necessary

File: route-rs-packets/src/types.rs
Comment (eeverett6) [inline route-rs-packets/src/types.rs]: experimentation

File: route-rs-packets/src/ethernet.rs
Comment (CollinValley) [reply route-rs-packets/src/ethernet.rs]: We chatted offline about this, my next push will shift us to just supporting EtherType instead of length field. I will still check that it is in a valid EtherType range.

File: route-rs-packets/src/types.rs
Comment (CollinValley) [reply route-rs-packets/src/types.rs]: whoops, fixed.

**Review by ddowl — APPROVED** [review]
lol I read this

**Pull request merged**
